File: HMCL/src/main/java/org/jackhuang/hmcl/ui/VersionSettingsController.java
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2018  huangyuhui <huanghongxun2008@126.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.ui;

import com.jfoenix.controls.*;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.control.ScrollPane;
import javafx.scene.control.Toggle;
import javafx.scene.image.Image;
import javafx.scene.layout.VBox;
import javafx.stage.FileChooser;
import org.jackhuang.hmcl.Main;
import org.jackhuang.hmcl.setting.EnumGameDirectory;
import org.jackhuang.hmcl.setting.Profile;
import org.jackhuang.hmcl.setting.VersionSetting;
import org.jackhuang.hmcl.task.Schedulers;
import org.jackhuang.hmcl.task.Task;
import org.jackhuang.hmcl.ui.construct.ComponentList;
import org.jackhuang.hmcl.ui.construct.ImagePickerItem;
import org.jackhuang.hmcl.ui.construct.MultiFileItem;
import org.jackhuang.hmcl.util.*;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.Collection;
import java.util.Optional;
import java.util.logging.Level;
import java.util.stream.Collectors;

public final class VersionSettingsController {
    private VersionSetting lastVersionSetting = null;
    private Profile profile;
    private String versionId;

    @FXML private VBox rootPane;
    @FXML private ScrollPane scroll;
    @FXML private JFXTextField txtWidth;
    @FXML private JFXTextField txtHeight;
    @FXML private JFXTextField txtMaxMemory;
    @FXML private JFXTextField txtJVMArgs;
    @FXML private JFXTextField txtGameArgs;
    @FXML private JFXTextField txtMetaspace;
    @FXML private JFXTextField txtWrapper;
    @FXML private JFXTextField txtPrecallingCommand;
    @FXML private JFXTextField txtServerIP;
    @FXML private ComponentList advancedSettingsPane;
    @FXML private JFXComboBox<?> cboLauncherVisibility;
    @FXML private JFXCheckBox chkFullscreen;
    @FXML private Label lblPhysicalMemory;
    @FXML private JFXToggleButton chkNoJVMArgs;
    @FXML private JFXToggleButton chkNoCommon;
    @FXML private JFXToggleButton chkNoGameCheck;
    @FXML private MultiFileItem globalItem;
    @FXML private MultiFileItem javaItem;
    @FXML private MultiFileItem gameDirItem;
    @FXML private JFXToggleButton chkShowLogs;
    @FXML private ImagePickerItem iconPickerItem;

    @FXML
    private void initialize() {
        lblPhysicalMemory.setText(Main.i18n("settings.physical_memory") + ": " + OperatingSystem.TOTAL_MEMORY + "MB");

        FXUtils.smoothScrolling(scroll);

        Task.of(variables -> variables.set("list", JavaVersion.getJREs().values().stream().map(javaVersion ->
                javaItem.createChildren(javaVersion.getVersion(), javaVersion.getBinary().getAbsolutePath(), javaVersion)
        ).collect(Collectors.toList()))).subscribe(Schedulers.javafx(), variables ->
                javaItem.loadChildren(variables.<Collection<Node>>get("list"))
        );

        if (OperatingSystem.CURRENT_OS == OperatingSystem.WINDOWS)
            javaItem.getExtensionFilters().add(new FileChooser.ExtensionFilter("Java", "java.exe", "javaw.exe"));

        gameDirItem.loadChildren(Arrays.asList(
                gameDirItem.createChildren(Main.i18n("settings.advanced.game_dir.default"), EnumGameDirectory.ROOT_FOLDER),
                gameDirItem.createChildren(Main.i18n("settings.advanced.game_dir.independent"), EnumGameDirectory.VERSION_FOLDER)
        ));

        globalItem.loadChildren(Arrays.asList(
                globalItem.createChildren(Main.i18n("settings.type.global"), true),
                globalItem.createChildren(Main.i18n("settings.type.special"), false)
        ));
    }

    public void loadVersionSetting(Profile profile, String versionId) {
        rootPane.getChildren().remove(advancedSettingsPane);

        this.profile = profile;
        this.versionId = versionId;

        VersionSetting versionSetting = profile.getVersionSetting(versionId);

        gameDirItem.setDisable(profile.getRepository().isModpack(versionId));
        globalItem.setDisable(profile.getRepository().isModpack(versionId));

        if (lastVersionSetting != null) {
            FXUtils.unbindInt(txtWidth, lastVersionSetting.widthProperty());
            FXUtils.unbindInt(txtHeight, lastVersionSetting.heightProperty());
            FXUtils.unbindInt(txtMaxMemory, lastVersionSetting.maxMemoryProperty());
            FXUtils.unbindString(javaItem.getTxtCustom(), lastVersionSetting.javaDirProperty());
            FXUtils.unbindString(gameDirItem.getTxtCustom(), lastVersionSetting.gameDirProperty());
            FXUtils.unbindString(txtJVMArgs, lastVersionSetting.javaArgsProperty());
            FXUtils.unbindString(txtGameArgs, lastVersionSetting.minecraftArgsProperty());
            FXUtils.unbindString(txtMetaspace, lastVersionSetting.permSizeProperty());
            FXUtils.unbindString(txtWrapper, lastVersionSetting.wrapperProperty());
            FXUtils.unbindString(txtPrecallingCommand, lastVersionSetting.preLaunchCommandProperty());
            FXUtils.unbindString(txtServerIP, lastVersionSetting.serverIpProperty());
            FXUtils.unbindBoolean(chkFullscreen, lastVersionSetting.fullscreenProperty());
            FXUtils.unbindBoolean(chkNoGameCheck, lastVersionSetting.notCheckGameProperty());
            FXUtils.unbindBoolean(chkNoCommon, lastVersionSetting.noCommonProperty());
            FXUtils.unbindBoolean(chkShowLogs, lastVersionSetting.showLogsProperty());
            FXUtils.unbindEnum(cboLauncherVisibility);
        }

        FXUtils.bindInt(txtWidth, versionSetting.widthProperty());
        FXUtils.bindInt(txtHeight, versionSetting.heightProperty());
        FXUtils.bindInt(txtMaxMemory, versionSetting.maxMemoryProperty());
        FXUtils.bindString(javaItem.getTxtCustom(), versionSetting.javaDirProperty());
        FXUtils.bindString(gameDirItem.getTxtCustom(), versionSetting.gameDirProperty());
        FXUtils.bindString(txtJVMArgs, versionSetting.javaArgsProperty());
        FXUtils.bindString(txtGameArgs, versionSetting.minecraftArgsProperty());
        FXUtils.bindString(txtMetaspace, versionSetting.permSizeProperty());
        FXUtils.bindString(txtWrapper, versionSetting.wrapperProperty());
        FXUtils.bindString(txtPrecallingCommand, versionSetting.preLaunchCommandProperty());
        FXUtils.bindString(txtServerIP, versionSetting.serverIpProperty());
        FXUtils.bindBoolean(chkFullscreen, versionSetting.fullscreenProperty());
        FXUtils.bindBoolean(chkNoGameCheck, versionSetting.notCheckGameProperty());
        FXUtils.bindBoolean(chkNoCommon, versionSetting.noCommonProperty());
        FXUtils.bindBoolean(chkShowLogs, versionSetting.showLogsProperty());
        FXUtils.bindEnum(cboLauncherVisibility, versionSetting.launcherVisibilityProperty());

        boolean flag = false;
        Toggle defaultToggle = null;
        for (Toggle toggle : javaItem.getGroup().getToggles()) {
            if (toggle.getUserData() == Lang.invoke(versionSetting::getJavaVersion)) {
                toggle.setSelected(true);
                flag = true;
            } else if (toggle.getUserData() == JavaVersion.fromCurrentEnvironment()) {
                defaultToggle = toggle;
            }
        }

        javaItem.setToggleSelectedListener(newValue -> {
            if (javaItem.isCustomToggle(newValue)) {
                versionSetting.setJava("Custom");
            } else {
                versionSetting.setJava(((JavaVersion) newValue.getUserData()).getVersion());
            }
        });

        if (!flag) {
            Optional.ofNullable(defaultToggle).ifPresent(t -> t.setSelected(true));
        }

        versionSetting.javaDirProperty().setChangedListener(it -> initJavaSubtitle(versionSetting));
        versionSetting.javaProperty().setChangedListener(it -> initJavaSubtitle(versionSetting));
        initJavaSubtitle(versionSetting);

        if (versionSetting.isUsesGlobal())
            globalItem.getGroup().getToggles().stream().filter(it -> it.getUserData() == Boolean.TRUE).findFirst().ifPresent(it -> it.setSelected(true));
        else
            globalItem.getGroup().getToggles().stream().filter(it -> it.getUserData() == Boolean.FALSE).findFirst().ifPresent(it -> it.setSelected(true));
        globalItem.setToggleSelectedListener(newValue -> {
            if ((Boolean) newValue.getUserData())
                profile.globalizeVersionSetting(versionId);
            else
                profile.specializeVersionSetting(versionId);

            Platform.runLater(() -> loadVersionSetting(profile, versionId));
        });

        versionSetting.usesGlobalProperty().setChangedListenerAndOperate(it ->
                globalItem.setSubtitle(Main.i18n(versionSetting.isUsesGlobal() ? "settings.type.global" : "settings.type.special")));

        gameDirItem.getGroup().getToggles().stream()
                .filter(it -> it.getUserData() == versionSetting.getGameDirType())
                .findFirst().ifPresent(toggle -> toggle.setSelected(true));

        gameDirItem.setCustomUserData(EnumGameDirectory.CUSTOM);
        gameDirItem.setToggleSelectedListener(newValue -> versionSetting.setGameDirType((EnumGameDirectory) newValue.getUserData()));

        versionSetting.gameDirProperty().setChangedListener(it -> initGameDirSubtitle(versionSetting));
        versionSetting.gameDirTypeProperty().setChangedListener(it -> initGameDirSubtitle(versionSetting));
        initGameDirSubtitle(versionSetting);

        lastVersionSetting = versionSetting;

        loadIcon();
    }

    private void initJavaSubtitle(VersionSetting versionSetting) {
        Task.of(variables -> variables.set("java", versionSetting.getJavaVersion()))
                .subscribe(Task.of(Schedulers.javafx(),
                        variables -> javaItem.setSubtitle(variables.<JavaVersion>getOptional("java")
                                .map(JavaVersion::getBinary).map(File::getAbsolutePath).orElse("Invalid Java Directory"))));
    }

    private void initGameDirSubtitle(VersionSetting versionSetting) {
        gameDirItem.setSubtitle(profile.getRepository().getRunDirectory(versionId).getAbsolutePath());
    }

    @FXML
    private void onShowAdvanced() {
        if (!rootPane.getChildren().contains(advancedSettingsPane))
            rootPane.getChildren().add(advancedSettingsPane);
        else
            rootPane.getChildren().remove(advancedSettingsPane);
    }

    @FXML
    private void onExploreIcon() {
        FileChooser chooser = new FileChooser();
        chooser.getExtensionFilters().add(new FileChooser.ExtensionFilter(Main.i18n("extension.png"), "*.png"));
        File selectedFile = chooser.showOpenDialog(Controllers.getStage());
        if (selectedFile != null) {
            File iconFile = profile.getRepository().getVersionIcon(versionId);
            try {
                FileUtils.copyFile(selectedFile, iconFile);
                loadIcon();
            } catch (IOException e) {
                Logging.LOG.log(Level.SEVERE, "Failed to copy icon file from " + selectedFile + " to " + iconFile, e);
            }
        }
    }

    private void loadIcon() {
        File iconFile = profile.getRepository().getVersionIcon(versionId);
        if (iconFile.exists())
            iconPickerItem.setImage(new Image("file:" + iconFile.getAbsolutePath()));
        else
            iconPickerItem.setImage(FXUtils.DEFAULT_ICON);
        FXUtils.limitSize(iconPickerItem.getImageView(), 32, 32);
    }
}

File: HMCL/src/main/kotlin/org/jackhuang/hmcl/ui/VersionSettingsController.kt
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2017  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.ui

import com.jfoenix.controls.*
import javafx.beans.InvalidationListener
import javafx.beans.binding.Bindings
import javafx.beans.value.ChangeListener
import javafx.fxml.FXML
import javafx.geometry.Pos
import javafx.scene.control.Label
import javafx.scene.control.ScrollPane
import javafx.scene.control.Toggle
import javafx.scene.control.ToggleGroup
import javafx.scene.layout.*
import javafx.stage.DirectoryChooser
import org.jackhuang.hmcl.i18n
import org.jackhuang.hmcl.setting.VersionSetting
import org.jackhuang.hmcl.ui.construct.ComponentList
import org.jackhuang.hmcl.ui.construct.NumberValidator
import org.jackhuang.hmcl.util.JavaVersion
import org.jackhuang.hmcl.util.OS

class VersionSettingsController {
    var lastVersionSetting: VersionSetting? = null
    @FXML lateinit var rootPane: VBox
    @FXML lateinit var scroll: ScrollPane
    @FXML lateinit var txtWidth: JFXTextField
    @FXML lateinit var txtHeight: JFXTextField
    @FXML lateinit var txtMaxMemory: JFXTextField
    @FXML lateinit var txtJVMArgs: JFXTextField
    @FXML lateinit var txtGameArgs: JFXTextField
    @FXML lateinit var txtMetaspace: JFXTextField
    @FXML lateinit var txtWrapper: JFXTextField
    @FXML lateinit var txtPrecallingCommand: JFXTextField
    @FXML lateinit var txtServerIP: JFXTextField
    @FXML lateinit var txtJavaDir: JFXTextField
    @FXML lateinit var advancedSettingsPane: ComponentList
    @FXML lateinit var cboLauncherVisibility: JFXComboBox<*>
    @FXML lateinit var cboRunDirectory: JFXComboBox<*>
    @FXML lateinit var chkFullscreen: JFXCheckBox
    @FXML lateinit var lblPhysicalMemory: Label
    @FXML lateinit var chkNoJVMArgs: JFXToggleButton
    @FXML lateinit var chkNoCommon: JFXToggleButton
    @FXML lateinit var chkNoGameCheck: JFXToggleButton
    @FXML lateinit var componentJava: ComponentList
    @FXML lateinit var javaPane: VBox
    @FXML lateinit var javaPaneCustom: BorderPane
    @FXML lateinit var radioCustom: JFXRadioButton
    @FXML lateinit var btnJavaSelect: JFXButton

    val javaGroup = ToggleGroup()

    fun initialize() {
        lblPhysicalMemory.text = i18n("settings.physical_memory") + ": ${OS.TOTAL_MEMORY}MB"

        scroll.smoothScrolling()

        val limit = 300.0
        //txtJavaDir.limitWidth(limit)
        txtMaxMemory.limitWidth(limit)
        cboLauncherVisibility.limitWidth(limit)
        cboRunDirectory.limitWidth(limit)

        val limitHeight = 10.0
        chkNoJVMArgs.limitHeight(limitHeight)
        chkNoCommon.limitHeight(limitHeight)
        chkNoGameCheck.limitHeight(limitHeight)

        fun validation(field: JFXTextField) = InvalidationListener { field.validate() }
        fun validator(nullable: Boolean = false) = NumberValidator(nullable).apply { message = "Must be a number." }

        txtWidth.setValidators(validator())
        txtWidth.textProperty().addListener(validation(txtWidth))
        txtHeight.setValidators(validator())
        txtHeight.textProperty().addListener(validation(txtHeight))
        txtMaxMemory.setValidators(validator())
        txtMaxMemory.textProperty().addListener(validation(txtMaxMemory))
        txtMetaspace.setValidators(validator(true))
        txtMetaspace.textProperty().addListener(validation(txtMetaspace))

        javaPane.children.clear()
        javaPane.children += createJavaPane(JavaVersion.fromCurrentEnvironment(), javaGroup)
        JavaVersion.JAVAS.values.forEach { javaVersion ->
            javaPane.children += createJavaPane(javaVersion, javaGroup)
        }
        javaPane.children += javaPaneCustom
        javaPaneCustom.limitHeight(20.0)
        radioCustom.toggleGroup = javaGroup
        txtJavaDir.disableProperty().bind(radioCustom.selectedProperty().not())
        btnJavaSelect.disableProperty().bind(radioCustom.selectedProperty().not())
    }

    private fun createJavaPane(java: JavaVersion, group: ToggleGroup): Pane {
        return BorderPane().apply {
            style = "-fx-padding: 3;"
            limitHeight(20.0)
            left = JFXRadioButton(java.longVersion).apply {
                toggleGroup = group
                userData = java
            }
            right = Label(java.binary.absolutePath).apply {
                styleClass += "subtitle-label"
                style += "-fx-font-size: 10;"
            }
        }
    }

    fun loadVersionSetting(version: VersionSetting) {
        rootPane.children -= advancedSettingsPane

        lastVersionSetting?.apply {
            widthProperty.unbind()
            heightProperty.unbind()
            maxMemoryProperty.unbind()
            javaArgsProperty.unbind()
            minecraftArgsProperty.unbind()
            permSizeProperty.unbind()
            wrapperProperty.unbind()
            precalledCommandProperty.unbind()
            serverIpProperty.unbind()
            fullscreenProperty.unbind()
            notCheckGameProperty.unbind()
            noCommonProperty.unbind()
            javaDirProperty.unbind()
            unbindEnum(cboLauncherVisibility)
            unbindEnum(cboRunDirectory)
        }

        bindInt(txtWidth, version.widthProperty)
        bindInt(txtHeight, version.heightProperty)
        bindInt(txtMaxMemory, version.maxMemoryProperty)
        bindString(txtJavaDir, version.javaDirProperty)
        bindString(txtJVMArgs, version.javaArgsProperty)
        bindString(txtGameArgs, version.minecraftArgsProperty)
        bindString(txtMetaspace, version.permSizeProperty)
        bindString(txtWrapper, version.wrapperProperty)
        bindString(txtPrecallingCommand, version.precalledCommandProperty)
        bindString(txtServerIP, version.serverIpProperty)
        bindEnum(cboLauncherVisibility, version.launcherVisibilityProperty)
        bindEnum(cboRunDirectory, version.gameDirTypeProperty)
        bindBoolean(chkFullscreen, version.fullscreenProperty)
        bindBoolean(chkNoGameCheck, version.notCheckGameProperty)
        bindBoolean(chkNoCommon, version.noCommonProperty)

        val javaGroupKey = "java_group.listener"
        @Suppress("UNCHECKED_CAST")
        if (javaGroup.properties.containsKey(javaGroupKey))
            javaGroup.selectedToggleProperty().removeListener(javaGroup.properties[javaGroupKey] as ChangeListener<in Toggle>)

        var flag = false
        var defaultToggle: JFXRadioButton? = null
        for (toggle in javaGroup.toggles)
            if (toggle is JFXRadioButton)
                if (toggle.userData == version.javaVersion) {
                    toggle.isSelected = true
                    flag = true
                } else if (toggle.userData == JavaVersion.fromCurrentEnvironment()) {
                    defaultToggle = toggle
                }

        val listener = ChangeListener<Toggle> { _, _, newValue ->
            if (newValue == radioCustom) { // Custom
                version.java = "Custom"
            } else {
                version.java = ((newValue as JFXRadioButton).userData as JavaVersion).longVersion
            }
        }
        javaGroup.properties[javaGroupKey] = listener
        javaGroup.selectedToggleProperty().addListener(listener)

        if (!flag) {
            defaultToggle?.isSelected = true
        }

        version.javaDirProperty.setChangedListener { initJavaSubtitle(version) }
        version.javaProperty.setChangedListener { initJavaSubtitle(version)}
        initJavaSubtitle(version)

        lastVersionSetting = version
    }

    private fun initJavaSubtitle(version: VersionSetting) {
        componentJava.subtitle = version.javaVersion?.binary?.absolutePath ?: "Invalid Java Directory"
    }

    fun onShowAdvanced() {
        if (!rootPane.children.contains(advancedSettingsPane))
            rootPane.children += advancedSettingsPane
        else
            rootPane.children.remove(advancedSettingsPane)
    }

    fun onExploreJavaDir() {
        val chooser = DirectoryChooser()
        chooser.title = i18n("settings.choose_javapath")
        val selectedDir = chooser.showDialog(Controllers.stage)
        if (selectedDir != null)
            txtJavaDir.text = selectedDir.absolutePath
    }
}